### PR TITLE
move param's device check to `_init_group` for fused

### PR DESCRIPTION
**Pull request by crcrpar**
There could be some cases where the params have the meta device when calling optimizer's dunder init and those params are materialized in the first computation. This change would allow such situation.

### CODE REVIEW

**Review by janeyx99 — COMMENTED** [review]
Is the purpose of this change to allow this to error for the functional call to optimizers? like `adam(...)`?

**Comment by crcrpar** [comment]
> Is the purpose of this change to allow this to error for the functional call to optimizers? like `adam(...)`?

I updated the description. In lightning-thunder we can delay the materialization of params to the first call of model which is usually after optimizer construction.

**Comment by janeyx99** [comment]
Why not create the optimizer only after the params get put on real devices?

**Comment by crcrpar** [comment]
it's possible but in lightning-thunder it's also possible to initialize a model on meta device, create an optimizer before materializing and converting the model into fully sharded

**Comment by janeyx99** [comment]
okay, I see, that is indeed reasonable to create the optimizer, do some shuffling, and then call the first step. However, I would anticipate that, in most cases, the state and params will not change after the state's been initialized. Given this, we don't want to run the device/dtype check every single step as that would be redundant/slow. So what if we move this check to under the first instantiation of state (under the `if len(state) == 0`)?

**Comment by janeyx99** [comment]
Also looping in @anijain2305 @mlazos as _init_group would change as a result of this PR and there may need to be dynamo changes. To summarize this change FYI: this moves the device/dtype check for fused optimizers from the constructor (typically not expected to be compiled) to _init_group (specially handled in compile).

File: torch/optim/adam.py [hunk 1]
```diff
@@ -127,8 +117,17 @@ def _init_group(
         state_steps,
     ):
         has_complex = False
+        fused_supported_devices = _get_fused_kernels_supported_devices()
```
Comment (janeyx99) [inline torch/optim/adam.py:120]: We should do a similar pattern for adagrad for consistency

**Review by mlazos — APPROVED** [review]
Moving error checks to `_init_group` shouldn't be an issue and should be good for code organization as well.

**Review by janeyx99 — COMMENTED** [review]
Add tests in test_optim.py (maybe through test_error_cases).

And add the same check in adagrad.py for consistency

File: test/test_optim.py [hunk 2]
```diff
@@ -1024,6 +1024,49 @@ def test_fused_matches_forloop(self, device, dtype, optim_info):
             self.skipTest("MPS supports only torch.float16 and torch.float32")
         self._test_derived_optimizers(device, dtype, optim_info, "fused")
 
+    @optims(
+        [optim for optim in optim_db if "fused" in optim.supported_impls],
+        dtypes=floating_types_and(
+            torch.bfloat16,
+            torch.float16,
```
Comment (janeyx99) [inline test/test_optim.py:1031]: we probably do not need to test more than one dtype, no?

File: torch/optim/adagrad.py [hunk 3]
```diff
@@ -136,6 +125,10 @@ def _init_group(self, group, params_with_grad, grads, state_sums, state_steps):
         has_sparse_grad, has_complex = False, False
         for p in group["params"]:
             if p.grad is not None:
+                if group["fused"] and getattr(
+                    self, "_need_device_dtype_check_for_fused", True
+                ):
+                    _device_dtype_check_for_fused(p, cuda_unsupported=True)
```
Comment (janeyx99) [inline torch/optim/adagrad.py:131]:
```suggestion
                    _device_dtype_check_for_fused(p, cuda_unsupported=True)
                    self._need_device_dtype_check_for_fused = False
```

File: torch/optim/adagrad.py [hunk 4]
```diff
@@ -68,21 +68,10 @@ def __init__(
...
-                for pg in self.param_groups
-                for p in pg["params"]
-            ):
-                raise RuntimeError(
-                    "`fused=True` requires all the params to be floating point Tensors of "
-                    f"supported devices: {fused_supported_devices}."
-                )
             if foreach:
                 raise RuntimeError("`fused` and `foreach` cannot be `True` together.")
+            self._step_supports_amp_scaling = True
```
Comment (janeyx99) [inline torch/optim/adagrad.py:73]: what does this do?

File: torch/optim/adagrad.py (hunk 3 above)
Comment (crcrpar) [reply torch/optim/adagrad.py:132]: thanks

File: torch/optim/adagrad.py (hunk 4 above)
Comment (crcrpar) [reply torch/optim/adagrad.py]: a line from wrong copy paste that I forgot to remove. thanks

File: test/test_optim.py (hunk 2 above)
Comment (crcrpar) [reply test/test_optim.py]: agreed. we don't need low prec cases

File: test/test_optim.py [hunk 5]
```diff
@@ -1024,6 +1024,46 @@ def test_fused_matches_forloop(self, device, dtype, optim_info):
...
+    def test_fused_error_on_params_on_meta(self, device, dtype, optim_info):
+        if _get_device_type(device) not in optim_info.supports_fused_on:
+            self.skipTest(
+                f"{device} is not supported for fused on {optim_info.optim_cls.__name__}"
+            )
+        if _get_device_type(device) == "mps" and dtype not in (
+            torch.float16,
+            torch.float32,
+        ):
+            self.skipTest("MPS supports only torch.float16 and torch.float32")
```
Comment (janeyx99) [inline test/test_optim.py:1040]:
```suggestion
```

**Review by janeyx99 — APPROVED** [review]

**Pull request closed without merging**